src/cursor/common.rs

```diff
@@ -214,3 +249,43 @@ pub(crate) struct CursorInformation {
     pub(crate) batch_size: Option<u32>,
     pub(crate) max_time: Option<Duration>,
 }
+
+#[derive(Clone, Debug)]
+pub(crate) struct PinnedConnection {
+    connection: Option<Arc<Mutex<Connection>>>,
+    is_valid: bool,
+}
+
+impl PinnedConnection {
+    pub(super) fn new(conn: Option<Connection>) -> Self {
+        Self {
+            connection: conn.map(|c| Arc::new(Mutex::new(c))),
+            is_valid: true,
+        }
+    }
+
+    pub(super) async fn lock(&self) -> Option<MutexGuard<'_, Connection>> {
+        match &self.connection {
+            Some(c) => Some(c.lock().await),
+            None => None,
+        }
+    }
+}
+
+pub(super) fn kill_cursor(
```

Comment:
I opted to make this a free function rather than a method on `PinnedConnection` because it doesn't feel like it belongs to the connection as such.